ArduCopter/events.cpp

```diff
@@ -4,152 +4,235 @@
  *       This event will be called when the failsafe changes
  *       boolean failsafe reflects the current state
  */
+
+bool Copter::failsafe_option(FailsafeOption opt) const
+{
+    return (g2.fs_options & (uint32_t)opt);
+}
+
 void Copter::failsafe_radio_on_event()
 {
-    // if motors are not armed there is nothing to do
-    if( !motors->armed() ) {
-        return;
-    }
+    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_OCCURRED);
 
+    int16_t action;
+    action = g.failsafe_throttle; //The action set in FS_THR_ENABLE
+
+    // Conditions to deviate from FS_THR_ENABLE selection and sent specigic GCS warning
     if (should_disarm_on_failsafe()) {
+        // should immediately disarm when we're on the ground
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - DISARMING");
         arming.disarm();
+        action = FS_THR_DISABLED;
+
+    } else if (flightmode->is_landing() && ((battery.has_failsafed() && battery.get_highest_failsafe_priority() <= FAILSAFE_LAND_PRIORITY))) {
+        // Allow landing to continue when battery failsafe requires it (not a user option)
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO + BATTERY FAILSAFE - CONTINUING LANDING");
+        action = FS_THR_ENABLED_ALWAYS_LAND;
+
+    } else if (flightmode->is_landing() && failsafe_option(FailsafeOption::THR_LAND_CONTINUE)) {
+        // Allow landing to continue when FS_OPTIONS is set to continue landing
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING LANDING");
+        action = FS_THR_ENABLED_ALWAYS_LAND;
+
+    } else if (control_mode == Mode::Number::AUTO && failsafe_option(FailsafeOption::THR_MISSION_CONTINUE)) {
+        // Allow mission to continue when FS_OPTIONS is set to continue mission
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING AUTO MODE MISSION");       
+        action = FS_THR_DISABLED;
+
+    } else if ((control_mode == Mode::Number::GUIDED || control_mode == Mode::Number::GUIDED_NOGPS) &&
```

Comment:
we should use the flightmode->in_guided_mode() instead of checking the modes directly

Reply:
Done